File: src/main/java/at/searles/parsing/Fold.java
package at.searles.parsing;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface Fold<T, U, V> {
    V apply(ParserStream stream, @NotNull T left, @NotNull U right); // must not return null

    default @Nullable
    T leftInverse(@NotNull V result) {
        return null;
    }

    default @Nullable
    U rightInverse(@NotNull V result) {
        return null;
    }
}

File: src/main/kotlin/at/searles/parsing/parser/Fold.kt
package at.searles.parsing.parser

fun interface Fold<A, B, C> {
    fun fold(left: A, right: B): C
    fun invertLeft(value: C): FnResult<A> = error("Not invertible function")
    fun invertRight(value: C): FnResult<B> = error("Not invertible function")
}